### PR TITLE
Sending Single Message Per Client Prefix/ User Prefix

### PR DESCRIPTION
**Pull request by gekco**
previously for each channel subscribed to a group message was  being pushed  to list stored at channel key. I have grouped those messages by sending list of channels associated with that that channel redis key with a single message  . It saves a great deal of time by reducing the number of push and pops from redis.
Fixes : #7, #83

### CODE REVIEW

File: channels_redis/core.py [hunk 1]
```diff
@@ -310,33 +315,34 @@ def _setup_encryption(self, symmetric_encryption_keys):
...
-                    end
-                end
-            """ % self.expiry
+                        for i=1,#KEYS do
+                            if redis.call('LLEN', KEYS[i]) < tonumber(ARGV[i + #KEYS]) then
+                                redis.call('RPUSH', KEYS[i], ARGV[i])
+                                redis.call('EXPIRE', KEYS[i], %d)
+                            end
+                        end
+                    """ % self.expiry
```
Comment (andrewgodwin) [inline channels_redis/core.py]: Why did you indent this without changing it?

File: channels_redis/core.py [hunk 2]
```diff
@@ -350,26 +356,38 @@ def _map_channel_to_connection(self, channel_names, message):
...
+                channel_key_to_message[channel_key] = message
+                channel_key_to_capacity[channel_key] = self.get_capacity(channel)
+            else:
+                channel_key_to_message[channel_key]["__asgi_channel__"].append(channel)
+
+        for key in channel_keys:
+            channel_key_to_message[key]["__asgi_channel__"] = ",".join(channel_key_to_message[key]["__asgi_channel__"])
+            channel_key_to_message[key] = self.serialize(channel_key_to_message[key])
+
+        return channel_keys, connection_to_channel_keys, channel_key_to_message, channel_key_to_capacity
```
Comment (andrewgodwin) [inline channels_redis/core.py]: This block needs some serious comments to say what's happening - I'm finding it hard to read.

File: channels_redis/core.py (hunk 1 above)
Comment (gekco) [reply channels_redis/core.py]: i guess the IDE did it. i will revert it.

File: channels_redis/core.py (hunk 2 above)
Comment (gekco) [reply channels_redis/core.py]: on it

**Comment by gekco** [comment]
(Correct me if i am understanding this wrong)
say there is a group name 'Group-A' with 100 members(channels)

originally when i called group_send('Group-A', message_dictionary) 
the same message was being pushed 100 times each time with a different __asgi_channel__ value   to the same  redis queue(because multiple channels have same channel_key ) which was being popped at function receive_single 100 times obviously and the message was getting processed for a single channel.

Modifications:
what i did was coalesce those messages that is for those 100 channels  i will be sending a single message for all the channels with the same channel key by sending a comma separated string of channels instead of a single channel. On the receiving end i just process the message for all of the channels that are received in the message. This Largely reduces the time to push group messages onto channel redis layer and the number of pops. 
For Example if all the 100 channels have the same channel key there will be a single message for all the 100 channels which will be pushed on the queue and popped from the queue.

File: channels_redis/core.py [hunk 3]
```diff
@@ -310,66 +316,102 @@ def _setup_encryption(self, symmetric_encryption_keys):
...
 
             # We need to send the capacity for each channel
-            args += [channel_to_capacity[channel_name] for channel_name in channel_names
-                    if channel_to_key[channel_name] in channel_redis_keys]
+            args += [channel_keys_to_capacity[channel_key] for channel_key in channel_keys
+                    if channel_key in channel_redis_keys]
 
+            # channel_keys does not contain a single redis key more than once
             async with self.connection(connection_index) as connection:
                 await connection.eval(group_send_lua, keys=channel_redis_keys, args=args)
```
Comment (gekco) [inline channels_redis/core.py]: previously channel_redis_keys would contain duplicate keys as multiple channels may map to same redis key. hence the more number of pushes.

**Comment by andrewgodwin** [comment]
OK, I mostly understand it now, but a couple of things I'd like to see changed:

- The `_map_channel_to_connection` function is now far too complex - that whole thing should be refactored so it doesn't have to pass around things as dicts of lists and so forth. Maybe flatten it more into the only place it's called.

- I'm not particularly happy with the comma-separated approach to the storing of channel names. Maybe just use a native msgpack list?

**Comment by gekco** [comment]
1. I have restructured my code a bit , also using a more relevant function name, eliminated some repitive code, flattened it, have a look. i am now using _map_channel_keys_to_connection function.
2. I am using a msgpack native list now

**Comment by andrewgodwin** [comment]
Thanks!

**Comment by gekco** [comment]
@andrewgodwin Thank you so much. This was my first PR to get merged . Yayy. :D

**Pull request merged**